pywr/_core.pyx

```diff
@@ -477,6 +477,14 @@ cdef class AbstractNode:
     cpdef double get_cost(self, ScenarioIndex scenario_index) except? -1:
         return 0.0
 
+    cpdef double[:] get_all_cost(self, double[:] out=None):
```

Comment:
Is there a specific reason for having the `out` arg? and does there need to be any check on its shape?

Reply:
It allows for a pre-allocated array to be given in which the values are set (i.e. avoids the need to reallocate space on every call). A solver would allocate space once per run to accept all data, and then pass those arrays as the `out` argument. This avoids an additional allocation and possibly a copy of the data.